File: log-analyzer/frontend/src/lib/schemas.ts
import { z } from 'zod'

// Enums
export const LogLevelSchema = z.enum(['DEBUG', 'INFO', 'WARNING', 'WARN', 'ERROR', 'CRITICAL', 'FATAL'])
export const LogFormatSchema = z.enum(['apache', 'nginx', 'syslog', 'json', 'generic'])
export const AnomalySeveritySchema = z.enum(['low', 'medium', 'high'])
export const AnomalyTypeSchema = z.enum([
  'error_spike',
  'unusual_pattern', 
  'frequency_anomaly',
  'time_anomaly',
  'ip_anomaly',
  'status_anomaly'
])

// Base Schemas
export const LogEntrySchema = z.object({
  line_number: z.number(),
  timestamp: z.string().nullable().optional(),
  level: LogLevelSchema,
  message: z.string(),
  service: z.string().nullable().optional(),
  format: LogFormatSchema,
  raw_line: z.string(),
  
  // Apache/Nginx specific fields
  ip_address: z.string().nullable().optional(),
  method: z.string().nullable().optional(),
  url: z.string().nullable().optional(),
  status_code: z.number().nullable().optional(),
  response_size: z.number().nullable().optional(),
  
  // Syslog specific fields
  hostname: z.string().nullable().optional(),
  
  // Metadata
  user_id: z.string().nullable().optional(),
  file_id: z.string().nullable().optional(),
  created_at: z.string().nullable().optional(),
  updated_at: z.string().nullable().optional(),
  
  // Analysis results
  anomaly_score: z.number().nullable().optional(),
  anomaly_type: AnomalyTypeSchema.nullable().optional(),
  tags: z.array(z.string()).nullable().optional(),
})

export const AnomalySchema = z.object({
  type: AnomalyTypeSchema,
  severity: AnomalySeveritySchema,
  description: z.string(),
  timestamp: z.string().nullable().optional(),
  details: z.record(z.any()),
})

export const AISummarySchema = z.object({
  summary: z.string(),
  insights: z.array(z.string()),
  recommendations: z.array(z.string()),
})

export const StatisticsSchema = z.object({
  error_count: z.number(),
  warning_count: z.number(),
  info_count: z.number(),
})

export const AnalysisResultSchema = z.object({
  total_entries: z.number(),
  anomalies: z.array(AnomalySchema),
  ai_summary: AISummarySchema.nullable().optional(),
  statistics: StatisticsSchema,
  timeline: z.record(z.record(z.number())).nullable().optional(),
})

export const TimelineDataSchema = z.object({
  timeline: z.record(z.record(z.number())),
  total_entries: z.number(),
})

export const FileInfoSchema = z.object({
  filename: z.string(),
  size: z.number(),
  uploaded_at: z.number(),
  path: z.string().nullable().optional(),
})

export const FileListResponseSchema = z.object({
  files: z.array(FileInfoSchema),
})

export const UploadResponseSchema = z.object({
  message: z.string(),
  filename: z.string(),
  file_path: z.string(),
  parse_result: z.array(z.record(z.any())).nullable().optional(),
})

export const DeleteFileResponseSchema = z.object({
  message: z.string(),
})

export const ErrorResponseSchema = z.object({
  error: z.string(),
  details: z.string().nullable().optional(),
})

export const SuccessResponseSchema = z.object({
  message: z.string(),
  data: z.record(z.any()).nullable().optional(),
})

// User Schemas
export const UserSchema = z.object({
  id: z.string(),
  email: z.string().email(),
  role: z.string(),
  provider: z.string(),
})

export const AuthResponseSchema = z.object({
  valid: z.boolean(),
  user: UserSchema.nullable().optional(),
  error: z.string().nullable().optional(),
})

// Request Schemas
export const FileUploadRequestSchema = z.object({
  file_path: z.string(),
  user_id: z.string().nullable().optional(),
})

export const AnalysisRequestSchema = z.object({
  file_path: z.string(),
})

export const AnomalyRequestSchema = z.object({
  file_path: z.string(),
})

export const TimelineRequestSchema = z.object({
  file_path: z.string(),
})

// Form Schemas
export const LoginFormSchema = z.object({
  email: z.string().email('Invalid email address'),
  password: z.string().min(6, 'Password must be at least 6 characters'),
})

export const SignUpFormSchema = z.object({
  email: z.string().email('Invalid email address'),
  password: z.string().min(6, 'Password must be at least 6 characters'),
  confirmPassword: z.string(),
}).refine((data) => data.password === data.confirmPassword, {
  message: "Passwords don't match",
  path: ["confirmPassword"],
})

// File Upload Schema
export const FileUploadSchema = z.object({
  files: z
    .custom<FileList>((val) => val instanceof FileList && val.length > 0, {
      message: 'Please select at least one file',
    }),
})


// API Response Schemas
export const APIResponseSchema = z.object({
  data: z.any().optional(),
  message: z.string().optional(),
  error: z.string().optional(),
})

// Type exports
export type LogLevel = z.infer<typeof LogLevelSchema>
export type LogFormat = z.infer<typeof LogFormatSchema>
export type AnomalySeverity = z.infer<typeof AnomalySeveritySchema>
export type AnomalyType = z.infer<typeof AnomalyTypeSchema>
export type LogEntry = z.infer<typeof LogEntrySchema>
export type Anomaly = z.infer<typeof AnomalySchema>
export type AISummary = z.infer<typeof AISummarySchema>
export type Statistics = z.infer<typeof StatisticsSchema>
export type AnalysisResult = z.infer<typeof AnalysisResultSchema>
export type TimelineData = z.infer<typeof TimelineDataSchema>
export type FileInfo = z.infer<typeof FileInfoSchema>
export type FileListResponse = z.infer<typeof FileListResponseSchema>
export type UploadResponse = z.infer<typeof UploadResponseSchema>
export type DeleteFileResponse = z.infer<typeof DeleteFileResponseSchema>
export type ErrorResponse = z.infer<typeof ErrorResponseSchema>
export type SuccessResponse = z.infer<typeof SuccessResponseSchema>
export type User = z.infer<typeof UserSchema>
export type AuthResponse = z.infer<typeof AuthResponseSchema>
export type FileUploadRequest = z.infer<typeof FileUploadRequestSchema>
export type AnalysisRequest = z.infer<typeof AnalysisRequestSchema>
export type AnomalyRequest = z.infer<typeof AnomalyRequestSchema>
export type TimelineRequest = z.infer<typeof TimelineRequestSchema>
export type LoginForm = z.infer<typeof LoginFormSchema>
export type SignUpForm = z.infer<typeof SignUpFormSchema>
export type FileUpload = z.infer<typeof FileUploadSchema>
export type APIResponse = z.infer<typeof APIResponseSchema>

// Validation helpers
export const validateLogEntry = (data: unknown): LogEntry => {
  return LogEntrySchema.parse(data)
}

export const validateAnalysisResult = (data: unknown): AnalysisResult => {
  return AnalysisResultSchema.parse(data)
}

export const validateFileList = (data: unknown): FileListResponse => {
  return FileListResponseSchema.parse(data)
}

export const validateUploadResponse = (data: unknown): UploadResponse => {
  return UploadResponseSchema.parse(data)
}

export const validateAuthResponse = (data: unknown): AuthResponse => {
  return AuthResponseSchema.parse(data)
}

// Safe parsing (returns null if invalid)
export const safeParse = <T>(schema: z.ZodSchema<T>, data: unknown): T | null => {
  try {
    return schema.parse(data)
  } catch {
    return null
  }
}

File: log-analyzer/backend/models/schemas.py
from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any, Union
from datetime import datetime
from enum import Enum

class LogLevel(str, Enum):
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    WARN = "WARN"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"
    FATAL = "FATAL"

class LogFormat(str, Enum):
    APACHE = "apache"
    NGINX = "nginx"
    SYSLOG = "syslog"
    JSON = "json"
    GENERIC = "generic"

class AnomalySeverity(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"

class AnomalyType(str, Enum):
    ERROR_SPIKE = "error_spike"
    UNUSUAL_PATTERN = "unusual_pattern"
    FREQUENCY_ANOMALY = "frequency_anomaly"
    TIME_ANOMALY = "time_anomaly"
    IP_ANOMALY = "ip_anomaly"
    STATUS_ANOMALY = "status_anomaly"

# Request Models
class FileUploadRequest(BaseModel):
    file_path: str = Field(..., description="Path to the uploaded file")
    user_id: Optional[str] = Field(None, description="User ID for the upload")

class AnalysisRequest(BaseModel):
    file_path: str = Field(..., description="Path to the log file to analyze")

class AnomalyRequest(BaseModel):
    file_path: str = Field(..., description="Path to the log file to analyze for anomalies")

class TimelineRequest(BaseModel):
    file_path: str = Field(..., description="Path to the log file for timeline analysis")

# Response Models
class LogEntryResponse(BaseModel):
    line_number: int = Field(..., description="Line number in the log file")
    timestamp: Optional[str] = Field(None, description="Parsed timestamp")
    level: LogLevel = Field(..., description="Log level")
    message: str = Field(..., description="Log message")
    service: Optional[str] = Field(None, description="Service name")
    format: LogFormat = Field(..., description="Log format type")
    raw_line: str = Field(..., description="Original log line")
    
    # Apache/Nginx specific fields
    ip_address: Optional[str] = Field(None, description="IP address")
    method: Optional[str] = Field(None, description="HTTP method")
    url: Optional[str] = Field(None, description="Request URL")
    status_code: Optional[int] = Field(None, description="HTTP status code")
    response_size: Optional[int] = Field(None, description="Response size in bytes")
    
    # Syslog specific fields
    hostname: Optional[str] = Field(None, description="Hostname")
    
    # Metadata
    user_id: Optional[str] = Field(None, description="User ID")
    file_id: Optional[str] = Field(None, description="File ID")
    created_at: Optional[datetime] = Field(None, description="Creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Last update timestamp")
    
    # Analysis results
    anomaly_score: Optional[float] = Field(None, description="Anomaly detection score")
    anomaly_type: Optional[AnomalyType] = Field(None, description="Type of anomaly detected")
    tags: Optional[List[str]] = Field(None, description="Tags associated with the log entry")

class AnomalyResponse(BaseModel):
    type: AnomalyType = Field(..., description="Type of anomaly")
    severity: AnomalySeverity = Field(..., description="Severity level")
    description: str = Field(..., description="Human-readable description")
    timestamp: Optional[str] = Field(None, description="When the anomaly occurred")
    details: Dict[str, Any] = Field(..., description="Detailed information about the anomaly")

class AISummaryResponse(BaseModel):
    summary: str = Field(..., description="AI-generated summary")
    insights: List[str] = Field(..., description="Key insights")
    recommendations: List[str] = Field(..., description="Recommended actions")

class StatisticsResponse(BaseModel):
    error_count: int = Field(..., description="Number of error logs")
    warning_count: int = Field(..., description="Number of warning logs")
    info_count: int = Field(..., description="Number of info logs")

class AnalysisResultResponse(BaseModel):
    total_entries: int = Field(..., description="Total number of log entries")
    anomalies: List[AnomalyResponse] = Field(..., description="Detected anomalies")
    ai_summary: Optional[AISummaryResponse] = Field(None, description="AI analysis summary")
    statistics: StatisticsResponse = Field(..., description="Log statistics")
    timeline: Optional[Dict[str, Dict[str, int]]] = Field(None, description="Timeline data grouped by hour")

class TimelineDataResponse(BaseModel):
    timeline: Dict[str, Dict[str, int]] = Field(..., description="Timeline data grouped by hour")
    total_entries: int = Field(..., description="Total number of entries")

class FileInfoResponse(BaseModel):
    filename: str = Field(..., description="File name")
    size: int = Field(..., description="File size in bytes")
    uploaded_at: float = Field(..., description="Upload timestamp")
    path: Optional[str] = Field(None, description="File path")

class FileListResponse(BaseModel):
    files: List[FileInfoResponse] = Field(..., description="List of uploaded files")

class UploadResponse(BaseModel):
    message: str = Field(..., description="Upload status message")
    filename: str = Field(..., description="Uploaded filename")
    file_path: str = Field(..., description="Path to uploaded file")
    parse_result: Optional[List[Dict[str, Any]]] = Field(None, description="Parsing result")

class DeleteFileResponse(BaseModel):
    message: str = Field(..., description="Deletion status message")

class ErrorResponse(BaseModel):
    error: str = Field(..., description="Error message")
    details: Optional[str] = Field(None, description="Additional error details")

class SuccessResponse(BaseModel):
    message: str = Field(..., description="Success message")
    data: Optional[Dict[str, Any]] = Field(None, description="Response data")

# User Models
class UserResponse(BaseModel):
    id: str = Field(..., description="User ID")
    email: str = Field(..., description="User email")
    role: str = Field(..., description="User role")
    provider: str = Field(..., description="Authentication provider")

class AuthResponse(BaseModel):
    valid: bool = Field(..., description="Token validity")
    user: Optional[UserResponse] = Field(None, description="User information")
    error: Optional[str] = Field(None, description="Error message if validation failed")

# Validation Models
class LogAnalysisConfig(BaseModel):
    max_log_size: int = Field(default=52428800, description="Maximum log size in bytes")
    max_file_size: int = Field(default=104857600, description="Maximum file size in bytes")
    allowed_extensions: List[str] = Field(default=[".log", ".txt", ".gz", ".tar"], description="Allowed file extensions")
    
    @validator('max_log_size', 'max_file_size')
    def validate_positive_size(cls, v):
        if v <= 0:
            raise ValueError('Size must be positive')
        return v

class AnomalyDetectionConfig(BaseModel):
    error_spike_threshold: float = Field(default=2.0, description="Standard deviation threshold for error spikes")
    frequency_anomaly_threshold: float = Field(default=2.0, description="Standard deviation threshold for frequency anomalies")
    business_hours_start: int = Field(default=9, description="Business hours start (0-23)")
    business_hours_end: int = Field(default=18, description="Business hours end (0-23)")
    suspicious_ip_threshold: int = Field(default=100, description="Minimum requests for IP anomaly detection")
    high_error_rate_threshold: float = Field(default=0.5, description="Error rate threshold for IP anomalies")
    
    @validator('business_hours_start', 'business_hours_end')
    def validate_hours(cls, v):
        if not 0 <= v <= 23:
            raise ValueError('Hour must be between 0 and 23')
        return v
    
    @validator('business_hours_end')
    def validate_business_hours(cls, v, values):
        if 'business_hours_start' in values and v <= values['business_hours_start']:
            raise ValueError('Business hours end must be after start')
        return v
